Reproducible NoHttpResponseException when making many requests

When hammering Nextcloud hard with requests (programmatically via DocumentsProvider) I reliably get a NoHttpResponseException which makes the request fail.

The log shows every few seconds:

```ruby
D/AdvancedSslSocketFactory: Creating SSL Socket with remote test.de:443, local null:0, params: org.apache.commons.httpclient.params.HttpConnectionParams@356164d
D/AdvancedSslSocketFactory:  ... with connection timeout 50000 and socket timeout 60000
```

Short before the failure it shows:

```ruby
D/AdvancedSslSocketFactory: Creating SSL Socket with remote test.de:443, local null:0, params: org.apache.commons.httpclient.params.HttpConnectionParams@612b1f
D/AdvancedSslSocketFactory:  ... with connection timeout 0 and socket timeout 0
```

Note that suddenly the timeouts are set to `0` which is strange.

Then eventually, the exception is thrown (which gets rethrown as a FileNotFoundException):

```ruby
D/OwnCloudClient #0: REQUEST PUT /remote.php/webdav/.SeedVaultAndroidBackup/1598533762174/kv/%40pm%40/bmV0LnRhbGVyLmNhc2hpZXI
E/DocumentsStorageProvider: RemoteOperationResult(mSuccess=false, mHttpCode=-1, mHttpPhrase=null, mException=org.apache.commons.httpclient.NoHttpResponseException: The server test.de failed to respond, mCode=UNKNOWN_ERROR, getLogMessage=Unrecovered transport exception)
E/DatabaseUtils: Writing exception to parcel
    android.os.ParcelableException: java.io.FileNotFoundException: Failed to upload document with path /.SeedVaultAndroidBackup/1598533762174/kv/@pm@/bmV0LnRhbGVyLmNhc2hpZXI
        at android.provider.DocumentsProvider.call(DocumentsProvider.java:1069)
        at android.content.ContentProvider.call(ContentProvider.java:2152)
        at android.content.ContentProvider$Transport.call(ContentProvider.java:477)
        at android.content.ContentProviderNative.onTransact(ContentProviderNative.java:277)
        at android.os.Binder.execTransactInternal(Binder.java:1021)
        at android.os.Binder.execTransact(Binder.java:994)
     Caused by: java.io.FileNotFoundException: Failed to upload document with path /.SeedVaultAndroidBackup/1598533762174/kv/@pm@/bmV0LnRhbGVyLmNhc2hpZXI
        at com.owncloud.android.providers.DocumentsStorageProvider.createFile(DocumentsStorageProvider.java:547)
        at com.owncloud.android.providers.DocumentsStorageProvider.createDocument(DocumentsStorageProvider.java:460)
        at android.provider.DocumentsProvider.callUnchecked(DocumentsProvider.java:1121)
        at android.provider.DocumentsProvider.call(DocumentsProvider.java:1067)
        at android.content.ContentProvider.call(ContentProvider.java:2152) 
```

Potentially related: #5940 as it has the same exception in the logs.
Further reading:
* https://williamsbdev.com/posts/no-http-response-exceptions/
* https://stackoverflow.com/a/10600762/4856311

--- a/src/main/java/com/owncloud/android/files/services/FileUploader.java
+++ b/src/main/java/com/owncloud/android/files/services/FileUploader.java
@@ -167,6 +167,11 @@ public class FileUploader extends Service
 
     public static final String KEY_LOCAL_BEHAVIOUR = "BEHAVIOUR";
 
+    /**
+     * Set to true if the HTTP library should disable automatic retries of uploads.
+     */
+    public static final String KEY_DISABLE_RETRIES = "DISABLE_RETRIES";
+
     public static final int LOCAL_BEHAVIOUR_COPY = 0;
     public static final int LOCAL_BEHAVIOUR_MOVE = 1;
     public static final int LOCAL_BEHAVIOUR_FORGET = 2;
@@ -407,6 +412,7 @@ public class FileUploader extends Service
         int localAction = intent.getIntExtra(KEY_LOCAL_BEHAVIOUR, LOCAL_BEHAVIOUR_FORGET);
         boolean isCreateRemoteFolder = intent.getBooleanExtra(KEY_CREATE_REMOTE_FOLDER, false);
         int createdBy = intent.getIntExtra(KEY_CREATED_BY, UploadFileOperation.CREATED_BY_USER);
+        boolean disableRetries = intent.getBooleanExtra(KEY_DISABLE_RETRIES, true);
         try {
             for (OCFile file : files) {
                 startNewUpload(
@@ -418,7 +424,8 @@ public class FileUploader extends Service
                     localAction,
                     isCreateRemoteFolder,
                     createdBy,
-                    file
+                    file,
+                    disableRetries
                 );
             }
         } catch (IllegalArgumentException e) {
@@ -446,7 +453,8 @@ public class FileUploader extends Service
         int localAction,
         boolean isCreateRemoteFolder,
         int createdBy,
-        OCFile file
+        OCFile file,
+        boolean disableRetries
     ) {
         OCUpload ocUpload = new OCUpload(file, user.toPlatformAccount());
         ocUpload.setFileSize(file.getFileLength());
@@ -469,7 +477,8 @@ public class FileUploader extends Service
             localAction,
             this,
             onWifiOnly,
-            whileChargingOnly
+            whileChargingOnly,
+            disableRetries
         );
         newUpload.setCreatedBy(createdBy);
         if (isCreateRemoteFolder) {
@@ -517,7 +526,8 @@ public class FileUploader extends Service
             upload.getLocalAction(),
             this,
             onWifiOnly,
-            whileChargingOnly
+            whileChargingOnly,
+            true
         );
 
         newUpload.addDataTransferProgressListener(this);
@@ -957,7 +967,7 @@ public class FileUploader extends Service
     }
 
     /**
-     * Upload and overwrite an already uploaded file
+     * Upload and overwrite an already uploaded file with disabled retries
      */
     public static void uploadUpdateFile(
         Context context,
@@ -966,7 +976,21 @@ public class FileUploader extends Service
         Integer behaviour,
         NameCollisionPolicy nameCollisionPolicy
     ) {
-        uploadUpdateFile(context, account, new OCFile[]{existingFile}, behaviour, nameCollisionPolicy);
+        uploadUpdateFile(context, account, new OCFile[]{existingFile}, behaviour, nameCollisionPolicy, true);
+    }
+
+    /**
+     * Upload and overwrite an already uploaded file
+     */
+    public static void uploadUpdateFile(
+        Context context,
+        Account account,
+        OCFile existingFile,
+        Integer behaviour,
+        NameCollisionPolicy nameCollisionPolicy,
+        boolean disableRetries
+    ) {
+        uploadUpdateFile(context, account, new OCFile[]{existingFile}, behaviour, nameCollisionPolicy, disableRetries);
     }
 
     /**
@@ -977,7 +1001,8 @@ public class FileUploader extends Service
         Account account,
         OCFile[] existingFiles,
         Integer behaviour,
-        NameCollisionPolicy nameCollisionPolicy
+        NameCollisionPolicy nameCollisionPolicy,
+        boolean disableRetries
     ) {
         Intent intent = new Intent(context, FileUploader.class);
 
@@ -985,6 +1010,7 @@ public class FileUploader extends Service
         intent.putExtra(FileUploader.KEY_FILE, existingFiles);
         intent.putExtra(FileUploader.KEY_LOCAL_BEHAVIOUR, behaviour);
         intent.putExtra(FileUploader.KEY_NAME_COLLISION_POLICY, nameCollisionPolicy);
+        intent.putExtra(FileUploader.KEY_DISABLE_RETRIES, disableRetries);
 
         if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
             context.startForegroundService(intent);

--- a/src/main/java/com/owncloud/android/operations/UploadFileOperation.java
+++ b/src/main/java/com/owncloud/android/operations/UploadFileOperation.java
@@ -121,6 +121,7 @@ public class UploadFileOperation extends SyncOperation {
     private boolean mOnWifiOnly;
     private boolean mWhileChargingOnly;
     private boolean mIgnoringPowerSaveMode;
+    private final boolean mDisableRetries;
 
     private boolean mWasRenamed;
     private long mOCUploadId;
@@ -184,8 +185,23 @@ public class UploadFileOperation extends SyncOperation {
                                int localBehaviour,
                                Context context,
                                boolean onWifiOnly,
-                               boolean whileChargingOnly
-    ) {
+                               boolean whileChargingOnly) {
+        this(uploadsStorageManager, connectivityService, powerManagementService, user, file, upload,
+             nameCollisionPolicy, localBehaviour, context, onWifiOnly, whileChargingOnly, true);
+    }
+
+    public UploadFileOperation(UploadsStorageManager uploadsStorageManager,
+                               ConnectivityService connectivityService,
+                               PowerManagementService powerManagementService,
+                               User user,
+                               OCFile file,
+                               OCUpload upload,
+                               FileUploader.NameCollisionPolicy nameCollisionPolicy,
+                               int localBehaviour,
+                               Context context,
+                               boolean onWifiOnly,
+                               boolean whileChargingOnly,
+                               boolean disableRetries) {
         if (upload == null) {
             throw new IllegalArgumentException("Illegal NULL file in UploadFileOperation creation");
         }
@@ -222,6 +238,7 @@ public class UploadFileOperation extends SyncOperation {
         // Ignore power save mode only if user explicitly created this upload
         mIgnoringPowerSaveMode = mCreatedBy == CREATED_BY_USER;
         mFolderUnlockToken = upload.getFolderUnlockToken();
+        mDisableRetries = disableRetries;
     }
 
     public boolean isWifiRequired() {
@@ -561,14 +578,18 @@ public class UploadFileOperation extends SyncOperation {
                                                                         mFile.getEtagInConflict(),
                                                                         timeStamp,
                                                                         onWifiConnection,
-                                                                        token);
+                                                                        token,
+                                                                        mDisableRetries
+                );
             } else {
                 mUploadOperation = new UploadFileRemoteOperation(encryptedTempFile.getAbsolutePath(),
                                                                  mFile.getParentRemotePath() + encryptedFileName,
                                                                  mFile.getMimeType(),
                                                                  mFile.getEtagInConflict(),
                                                                  timeStamp,
-                                                                 token);
+                                                                 token,
+                                                                 mDisableRetries
+                );
             }
 
             for (OnDatatransferProgressListener mDataTransferListener : mDataTransferListeners) {
@@ -796,13 +817,15 @@ public class UploadFileOperation extends SyncOperation {
                                                                         mFile.getMimeType(),
                                                                         mFile.getEtagInConflict(),
                                                                         timeStamp,
-                                                                        onWifiConnection);
+                                                                        onWifiConnection,
+                                                                        mDisableRetries);
             } else {
                 mUploadOperation = new UploadFileRemoteOperation(mFile.getStoragePath(),
                                                                  mFile.getRemotePath(),
                                                                  mFile.getMimeType(),
                                                                  mFile.getEtagInConflict(),
-                                                                 timeStamp);
+                                                                 timeStamp,
+                                                                 mDisableRetries);
             }
 
             for (OnDatatransferProgressListener mDataTransferListener : mDataTransferListeners) {

--- a/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.java
+++ b/src/main/java/com/owncloud/android/providers/DocumentsStorageProvider.java
@@ -249,8 +249,8 @@ public class DocumentsStorageProvider extends DocumentsProvider {
                             account,
                             ocFile,
                             LOCAL_BEHAVIOUR_MOVE,
-                            NameCollisionPolicy.OVERWRITE
-                                                     );
+                            NameCollisionPolicy.OVERWRITE,
+                            false);
                     } else { // error, no upload needed
                         Log_OC.e(TAG, "File was closed with an error: " + ocFile.getFileName(), error);
                     }
@@ -537,7 +537,8 @@ public class DocumentsStorageProvider extends DocumentsProvider {
                                                                      newFilePath,
                                                                      mimeType,
                                                                      "",
-                                                                     String.valueOf(System.currentTimeMillis() / 1000))
+                                                                     String.valueOf(System.currentTimeMillis() / 1000),
+                                                                     false)
             .execute(client);
 
         if (!result.isSuccess()) {

